### PR TITLE
Bluespace Closet

### PR DESCRIPTION
**Pull request by Jazz23**
**What does this PR do:**
This is the bluespace closet. If you thought you couldn't get any more pointless, here's an upgraded **closet**. It has a storage capacity of 60 (30 is the regular closet) and it can pass through objects/mobs (except for walls and doors). It's has a sweet looking animated sprite that I take 0 credit for (thanks to my friend Plausible). It can be printed from the protolathe with a req_tech of` "engineering" = 4, "programming" = 5, "bluespace" = 5, "magnets" = 4, "plasmatech" = 3`, and a materials list of `MAT_METAL = 5000, MAT_PLASMA = 2500, MAT_TITANIUM = 500, MAT_BLUESPACE = 500`.

The reason I made this was because I absolutely loose my mind when people bump into my locker full of bodies as I'm running through the halls tryna space them. Since mobs can walk through this closet, the convenience factor increases tenfold. Not only that, you can drag it over tables and grab the stuff off of tables for easier, faster item transportation. You **can** drag it over basic barricades, meaning you could stuff your EOC buddy inside and get through, but honestly its quicker to just deconstruct or break them rather than go to RnD and print a bluespace locker and stuff your friend inside and drag it.... yea you get the point. It also has transparency while moving or while an object with density is on top of it (table, people, etc). If there is any other things it should not be able to phase through (besides walls and doors) it's as simple as adding it to checklist in the move proc. Also feel free to modify the req tech or materials because I totally just winged that.

**Images of sprite/map changes (IF APPLICABLE):**

3mbha6f3r0
ztrlwa4avw
amdow_sma52onwad

**Changelog:**
:cl:
add: Bluespace Closet
imageadd: Bluespace closet animations/sprites
/:cl:

### CODE REVIEW

**Comment by AffectedArc07** [comment]
I like the concept but I feelthe sprite could do with a bit of work, mainly the inner strobing part.

**Comment by Superhats** [comment]
Will this take me to Narnia?

**Comment by variableundefined** [comment]
What is the point of the bluespace closet? Does it serve any role?

**Comment by Ralta** [comment]
I both love and hate this. I love the idea for 'legitimate' purposes (general transport and storage), but can see this getting abused for shenanigans since it passes through mobs, tables, barricades etc.

**Comment by KasparoVy** [comment]
Looks bloody awesome.

**Comment by davidchan13** [comment]
Are there any plans for malfunctions or accidents for if someone were to toss a bag of holding or similiar item into the closet and close it?

**Comment by Jazz23** [comment]
So funny thing, I also agreed with you guys about the open sprite and had my friend made the up and down motion as well. I tried both and went with the down only, I'll definitely change it and edit the spot to see what yall think. Also for the record it's not that glitchy in game, I just sharex'd it and got luck with a quasi seemleses loop for my PR. And about narnia, I actually was considering a teleport thing but that would be too much. 

And in regards to 
> I both love and hate this. I love the idea for 'legitimate' purposes (general transport and storage), but can see this getting abused for shenanigans since it passes through mobs, tables, barricades etc.
The exlusion list of things it can't phase through is as simple as adding the atom type in the if check. If this gets implemented, if an admin or player finds a way to abuse this, it is no trouble at all to quick patch it. Furthermore, if it comes down to it I could also make it a whitelist of items it can phase through, which kinda makes it a little less interesting since it might have unexpected behavior with stuff not on the whitelist. The point is, if it becomes a problem is really easy to nerf it.

**Comment by taukausanake** [comment]
I think the inside glow could be a little more smooth. Looks kind of thick right now and not as nice as it could.

I love this idea though. I was more thinking it would be like a TARDIS with a walk in space even though that would be a little dumb but the way this is designed is still good.

We need to bluespace all the things now

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 1]
```diff
@@ -422,4 +422,44 @@
...
+
+/obj/structure/closet/bluespace/CheckExit(atom/movable/AM)
+	IsAlone(AM, loc)
+	return TRUE
+
+/obj/structure/closet/bluespace/proc/IsAlone(atom/movable/AM, atom/location)
+	var/trans = FALSE
+	for(var/atom/A in location)
+		if(A.density && A != src && A != AM)
+			trans = TRUE
```
[the remark this answers is not in the document]
Comment (Jazz23) [reply code/game/objects/structures/crates_lockers/closets.dm]: I usually do totally forgot

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 2]
```diff
@@ -115,7 +115,7 @@
 		playsound(loc, sound, 15, 1, -3)
 	else
 		playsound(loc, 'sound/machines/click.ogg', 15, 1, -3)
-	density = 1
+	density = !istype(src, /obj/structure/closet/bluespace)
```
[the remark this answers is not in the document]
Comment (Jazz23) [reply code/game/objects/structures/crates_lockers/closets.dm]: Yea but wouldn't that override all the other code that goes along with closing? Or does it fire anyways...

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 3]
```diff
@@ -422,4 +422,44 @@
...
+			trans = TRUE
+	icon_opened = trans ? "bluespaceopentrans" : "bluespaceopen"
+	icon_closed = trans ? "bluespacetrans" : "bluespace"
+	icon_state = opened ? icon_opened : icon_closed
+	return TRUE
+
+/obj/structure/closet/bluespace/Crossed(atom/movable/AM)
+	if(AM.density)
+		icon_state = opened ? "bluespaceopentrans" : "bluespacetrans"
+	return
```
[the remark this answers is not in the document]
Comment (Jazz23) [reply code/game/objects/structures/crates_lockers/closets.dm]: I saw it somewhere else in the code so I thought I'd follow the pattern

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 4]
```diff
@@ -422,4 +422,44 @@
...
+
+/obj/structure/closet/bluespace
+	name = "bluespace closet"
+	desc = "A storage unit that moves and stores through the fourth dimension."
+	density = 0
+	icon_state = "bluespace"
+	icon_closed = "bluespace"
+	icon_opened = "bluespaceopen"
+	storage_capacity = 60
+	var/list/materials = list(MAT_METAL = 5000, MAT_PLASMA = 2500, MAT_TITANIUM = 500, MAT_BLUESPACE = 500)
```
[the remark this answers is not in the document]
Comment (Jazz23) [reply code/game/objects/structures/crates_lockers/closets.dm]: Again just copying what others have done, just trying to keep consistent the conventions of the other developers.

File: code/game/objects/structures/crates_lockers/closets.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jazz23) [reply code/game/objects/structures/crates_lockers/closets.dm]: Yea just tested it, it overrides the original /obj/structure/closet/close(). Am I misinterpreting what your saying?

**Comment by Jazz23** [comment]
> Are there any plans for malfunctions or accidents for if someone were to toss a bag of holding or similiar item into the closet and close it?

Haven't really thought about it, possibly.

File: code/game/objects/structures/crates_lockers/closets.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jazz23) [reply code/game/objects/structures/crates_lockers/closets.dm]: woops didn't know what . = ..() does...

**Comment by KasparoVy** [comment]
I really dug the scanline looking pattern personally, oh well.

File: code/game/objects/structures/crates_lockers/closets.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (Citinited) [reply code/game/objects/structures/crates_lockers/closets.dm]: Yeah `..()` is your friend for everything where you need to override something on a child 😄

**Comment by Jazz23** [comment]
> I really dug the scanline looking pattern personally, oh well. RIP the closet rave

hr0k7avbuh

How about dem apples

**Comment by AffectedArc07** [comment]
The front doesnt see  to be pulsing?

**Comment by Jazz23** [comment]
> The front doesnt see to be pulsing?

What?

**Comment by AffectedArc07** [comment]
Before
djz0po9wwu

After
jyvqg1m5mf

**Comment by Jazz23** [comment]
Oh shoot forgot thanks

**Comment by taukausanake** [comment]
Yeah the scan line is good and what you got now is what I was talking about. Still think it can be _better_ but I'm not going to complain with what you got now. You could make it a little faster like the original one though but with the same thinness/smoothness

**Comment by Jazz23** [comment]
> I really like this idea because I'm seeing a clear use for it as an alternative to the paramedic ambulance. While I use the ambulance I also am willing to admit it's a giant ass buttpain, gets moved around by fastmos too easily. And can't be walked through and can only fit one person on the bed, which regularly gets detached from the ambulance.
> 
> .. honestly would be nice to see a medical themed type one if it goes through that could be usable for that.

Bluespace Ambulance Cart? On it.

**Comment by variableundefined** [comment]
Conflicted with Magistrate's PR it seems. DM / Ping me ASAP when it is deconflicted so I can merge it.

**Pull request merged**